lib/src/main/java/com/diffplug/spotless/Lint.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright 2022-2024 DiffPlug
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.diffplug.spotless;
+
+import java.io.Serializable;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Models a linted line or line range. Note that there is no concept of severity level - responsibility
+ * for severity and confidence are pushed down to the configuration of the lint tool. If a lint makes it
+ * to Spotless, then it is by definition.
+ */
+public final class Lint implements Serializable {
+	/** Any exception which implements this interface will have its lints extracted and reported cleanly to the user. */
+	public interface Has {
+		List<Lint> getLints();
+	}
+
+	/** An exception for shortcutting execution to report a lint to the user. */
+	public static class ShortcutException extends RuntimeException implements Has {
+		public ShortcutException(Lint... lints) {
+			this(Arrays.asList(lints));
+		}
+
+		private final List<Lint> lints;
+
+		public ShortcutException(Collection<Lint> lints) {
+			this.lints = List.copyOf(lints);
+		}
+
+		@Override
+		public List<Lint> getLints() {
+			return lints;
+		}
+	}
+
+	private static final long serialVersionUID = 1L;
+
+	private int lineStart, lineEnd; // 1-indexed, inclusive
+	private String code; // e.g. CN_IDIOM https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#cn-class-implements-cloneable-but-does-not-define-or-use-clone-method-cn-idiom
+	private String msg;
+
+	private Lint(int lineStart, int lineEnd, String lintCode, String lintMsg) {
+		this.lineStart = lineStart;
+		this.lineEnd = lineEnd;
+		this.code = LineEnding.toUnix(lintCode);
+		this.msg = LineEnding.toUnix(lintMsg);
+	}
+
+	public static Lint create(String code, String msg, int lineStart, int lineEnd) {
+		if (lineEnd < lineStart) {
```

Comment:
Given `lineNumberFor`, `lineNumber=-1` could be given a specific semantic (e.g. `somewhere_undefined`).

Reply:
That's a good call, it's probably a mistake to pretend that every Lint knows the range that it applies to.

Reply:
addressed in [43cd94b](https://github.com/diffplug/spotless/pull/2149/commits/43cd94b5b577696311ccb71010fa015e10b59f90)